src/Token.php

```diff
@@ -37,6 +40,18 @@ public function value(): string
         return $this->value;
     }
 
+    public function valueUpper(): string
```

Comment:
i'm really not sure this method should be here... this is just an implementation detail for your tokenizer implementation. i think that a `strtoupper` was fine

Reply:
This has currently small (1.5%) speedup, I personally like this improvement as the code is then simpler and faster. As the `Token` is more or less internal class, I do not see any problem with this well defined and handly method.

Reply:
Do we even need to invoke `strtoupper($token->value())` all the time instead of using a variable? It would make `valueUpper()` obsolete.

Reply:
We need - as we need upper case prev/next token too.

Reply:
The amount of the previous token's value is manageable. Less method and functions invocations would also be in the interest of your previous reasoning.

Reply:
I have reverted the `Token::valueUpper()` improvement for now as it is unrelated to the main optimization doen by this PR.